smartlearner/interfaces/loss.py

```diff
@@ -5,20 +6,29 @@ class Loss(object):
     def __init__(self, model, dataset):
         self.model = model
         self.dataset = dataset
-        self.target = dataset.symb_targets
         self.consider_constant = []  # Part of the computational graph to be considered as a constant.
+        self.graph_updates = OrderedDict()
+        self._gradients = None
+
+    @property
+    def target(self):
+        return self.dataset.symb_targets
 
     def get_graph_output(self):
         output, updates = self.model.get_model_output(self.dataset.symb_inputs)
         return self._loss_function(output), updates
 
-    def get_gradients(self):
-        cost, updates = self.get_graph_output()
-        gparams = T.grad(cost=cost,
-                         wrt=self.model.parameters,
-                         consider_constant=self.consider_constant)
-        gradients = dict(zip(self.model.parameters, gparams))
-        return gradients, updates
+    @property
+    def gradients(self):
+        if self._gradients is None:
+            cost, updates_from_model_output = self.get_graph_output()
+            gparams = T.grad(cost=cost,
+                             wrt=self.model.parameters,
+                             consider_constant=self.consider_constant)
+            self._gradients = dict(zip(self.model.parameters, gparams))
+            self.graph_updates.update(updates_from_model_output)
+
+        return self._gradients
```

Comment:
Mostly, force feeding the updates through this tuple returns ensures the user don't forget about the updates.
The bestest would be for the objects to register their updates themselves to an update_aggregator.

It's probably one of those time I'd use module variables to keep track of them. It's not pretty, but the only other idea I have would be to attach them to the graph itself. I'm really not a big fan of moving them around, it's too error prone.